New `ThreeWChart` class for easy and interactive data visualization

yantavares opened the pull request:
## Description:

**This PR was made as a follow up to #122 .**

Enhancements Made Since the Initial PR:

- Added the `ThreeWChart` class to the `misc.py` module.
- Exposed the class in the `__init__.py` file.
- Revised the documentation to follow to **3W** guidelines.
- Updated the data loading process to use the 3W Toolkit's `load_instance` method.
- Created the `overviews/yantavres/main.ipynb` notebook to provide an overview of the class.
- Implemented general improvements to the chart styling.

---

This PR introduces a new `ThreeWChart` class to the `3W` repository, enabling the creation of interactive visualizations from 3W dataset files using Plotly. The `ThreeWChart` module provides an intuitive way to display time-series data with background shapes indicating key class transitions, along with a custom legend for better event understanding.

---

## Code Overview:
- **`ThreeWChart` class**:
  - **Attributes**:
    - `class_mapping`: Maps class IDs to descriptive event names.
    - `class_colors`: Assigns colors to each class for visual distinction.
  - **Methods**:
    - `__init__`: Initializes the class with parameters: `file_path`, `title`, `y_axis`, `use_dropdown`, and `dropdown_position`.
    - `_generate_class_mapping`: Generates a combined mapping of event labels (including transient states) to their descriptions.
    - `_generate_class_colors`: Automatically generates a color mapping for event labels, with lower opacity for transient states.
    - `_load_data`: Loads and preprocesses the dataset from a Parquet file using `load_instance`.
    - `_get_non_zero_columns`: Returns the list of columns that are not all zeros or NaN.
    - `_get_background_shapes`: Generates background shapes based on class transitions.
    - `_add_custom_legend`: Adds a custom legend with event names and colors to the chart.
    - `plot`: Generates the interactive chart with class annotations and displays it using Plotly.

---

## Example Usage:

- An overview on how to use the `ThreeWChart` class was added in `overviews/yantavares/main.ipynb`. Click [here](https://github.com/yantavares/3W/blob/feat/ThreeWChart/overviews/yantavares/main.ipynb) to see it.

---

## Dependencies:
- **Pandas**: For data manipulation and preprocessing.
- **Plotly**: For generating interactive visualizations.

yantavares commented:
I noticed that I used the wrong formatter in the first version, it is now fixed by commit **e4c4072**

ricardoevvargas commented:
Hello, @yantavares.

This PR is much better than PR #122. I noticed that all the adjustments listed in PR #122 have been made. Very good! 

However, additional (and simpler!) adjustments are important. We can implement them together. All you have to do is indicate which ones you would prefer not to implement.

1. The correct spelling is 3W **D**ataset, not 3W **d**ataset : - ))
2. In `overviews/yantavares/main.ipynb`, it is important to replace `3WChart` with `ThreeWChart` to avoid confusion;
3. The `class_mapping` attribute seems to be unnecessary, as the `LABELS_DESCRIPTIONS` constant already exists (generated from `dataset.ini`). We can keep this attribute, but updates to the 3W Dataset may require adjustments to the `ThreeWChart` class;
4. The term `Class Events` is not appropriate. Better to replace it with `Legend` (or something equivalent);
5. What you're calling a `class event` is actually a `variable`. A `class event` is something completely different. This adjustment is very important to avoid misunderstandings;
6. Selection/de-selection of certain legend objects is not resulting in changes to the graphic. Is that what is expected? Some users may think that this is a bug. Evaluate what can be done;
7. The `_get_non_zero_columns` method is not mentioned in the PR. As you've listed the others, it doesn't cost much to include it here;
8. You forgot to update the usage example in the PR. How about including a link to the overview you created? : - ))
9. It's better to use the `other_improvements` branch, as this PR contains contributions other than cosmetic ones. I'll make this adjustment right away.

Once again, thank you for submitting this PR.

yantavares commented:
Hello, @ricardoevvargas.

I apologize for the mistakes in the previous PR and thank you for the feedback and collaboration! I have made some changes, but I still have doubts about the best way to implement the functionality.

---

_The class_mapping attribute seems to be unnecessary, as the LABELS_DESCRIPTIONS constant already exists (generated from dataset.ini). We can keep this attribute, but updates to the 3W Dataset may require adjustments to the ThreeWChart class; :question:_

I updated the code to load all events from `dataset.ini`. However, the transient states are not present in the `dataset.ini` files, and I must add them manually for them to appear in the legend. My first approach was to add transient states **only** if `transient = True` was present in the event.

```py
    def _generate_class_mapping(self) -> Dict[int, str]:
        """
        Generate a combined mapping of event labels (including transient states) to their descriptions.
        This dynamically updates if an event's information changes in the configuration.
        """
        mapping = {}
        for event in EVENT_NAMES:
            section = DATASET_INI.get(event)
            base_label = section.getint("LABEL")
            description = section.get("DESCRIPTION")
            mapping[base_label] = description
            # If the event is transient, add an entry for its transient state.
            if section.getboolean("TRANSIENT"):
                mapping[base_label + TRANSIENT_OFFSET] = "Transient: " + description
        return mapping
```

This approach requires importing `DATASET_INI` and using its getters to access the configuration, which increases the class’s dependencies.

The alternative approach (which I chose) was to add the transient state to the `class_mapping` unconditionally, even if it does not exist in the dataset:

```py
    def _generate_class_mapping(self) -> Dict[int, str]:
        """Generate a combined mapping of event labels (including transient states) to their descriptions.

        Returns:
            Dict[int, str]: Mapping of event labels to their descriptions.
        """
        mapping = {}
        for label, description in LABELS_DESCRIPTIONS.items():
            mapping[label] = description
            mapping[label + TRANSIENT_OFFSET] = "Transient: " + description
        return mapping
```

This approach requires only one extra import, `TRANSIENT_OFFSET`, and avoids additional configuration lookups.  
Which approach is better? Do you have a better solution to the transient states mapping?

---

_Selection/de-selection of certain legend objects is not resulting in changes to the graphic. Is that what is expected? Some users may think that this is a bug. Evaluate what can be done; :question:_

The issue occurs because background shapes do not appear in the legend. I added the `_add_custom_legend` method to display them, but the toggle on/off functionality remains unsupported.

I evaluated several solutions. First, I attempted to use JavaScript to modify Plotly's behavior, but that added significant complexity. Next, I converted the background shapes to traces, similar to the curve, but the result did not work as expected. I then used checkboxes to mimic toggling of background shapes; however, this approach did not function correctly in the ipynb environment.

I therefore removed the toggle button behavior for the lines. Since there is only one line in the chart, toggling does not add significant contribution, and this avoids extra complexity. Do you have a better fix?

---

I updated the overview text, although further modifications are probably needed. Could you please help me to improve the text?

Please, feel free to make any changes necessary to the code as well as request further improvements. Thank you for helping make my contribution possible. :smile:

ricardoevvargas commented:
Hello, @yantavares.

I've noticed a great evolution in this PR. Very good! 👏👏👏

* I believe that the following adjustments are well implemented: 1, 2, 4, 5, 6, 7, and 9 👍
* I think you have implemented a suitable solution for the 6th adjustment 👍
* The 8th is pending, but it is quite simple: just edit the **Example Usage** section at the top of this PR. My suggestion is that you replace the code (the `chart` folder no longer exists) and the chart (outdated) with a sentence with a link to the overview you've created;
* Regarding the `class_mapping` attribute, I have a proposal. Replace the `_generate_class_mapping` method with a merge of two dicts. Which dicts am I talking about? The `LABELS_DESCRIPTIONS` constant that already exist and let's say a additional constant called `TRANSIENT_LABELS_DESCRIPTIONS` created by you in `base.py` (inside the if on line 90). The main advantage is that this new constant will be available to everyone, which will be an additional contribution from you. In this case, don't forget to expose this new constant in `__init__.py` (import from `base.py`). What do you think of this proposal?

Once again, thank you for submitting this PR.

yantavares commented:
Hello @ricardoevvargas,

I really appreciate your suggestion and have already implemented it in commits a67fc42 and b5f9aeb. The `TRANSIENT_LABELS_DESCRIPTIONS` constant is now used in `misc.py`. Additionally, I updated the PR description to link directly to the overview I created, providing a clearer explanation of the functionality.

Thank you again for helping me make this PR possible.
